### PR TITLE
implemented additionally functionality of formatters_col in to_latex

### LINKED ISSUE #26278: Feature request: easier digit format in to_latex()
#### Easier  digit format
It is possible to change the digit format for the columns in a latex table like this
```python
import pandas as pd
data  = [[1,2,3],
         [4,5,6],
         [7,8,9.001]]
df = pd.DataFrame(data, columns=['a', 'b', 'c'], index = ["foo", "bar" , "ho"])
cm0 =lambda x:'%1i'   %x
cm1 =lambda x:'%1.1f' %x
cm3= lambda x:'%1.3f' %x
format= [cm0, cm1, cm3]
latex_tab = df.to_latex( formatters= format)
print(latex_tab)

```
#### However it is really inconvenient to define everytime a new lambda function. It would be great if there would be the option:
```python
latex_tab = df.to_latex( data, formatters_col= ['%1i','%1.1f','%1.3f'])
```
Furthermore, changing the digits for one row is even more difficult. It would be great to have also the option 
```python
latex_tab = df.to_latex( data, formatters_index= ['%1i','%1.1f','%1.3f'])
```

### PR DESCRIPTION
**Pull request by wgranados**
- [ x] closes #26278
- [ x] tests added / passed
- [ x ] passes `black pandas`
- [ x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ x] whatsnew entry

Was having some issues rebasing on https://github.com/pandas-dev/pandas/pull/32666 so making a new PR with same changes. Closing other PR as well.

### CODE REVIEW

**Comment by arw2019** [comment]
cc @WillAyd @mroeschke

File: doc/source/whatsnew/v1.2.0.rst [hunk 1]
```diff
@@ -547,6 +547,17 @@ ExtensionArray
 - Fixed an inconsistency in :class:`PeriodArray`'s ``__init__`` signature to those of :class:`DatetimeArray` and :class:`TimedeltaArray` (:issue:`37289`)
 - Reductions for :class:`BooleanArray`, :class:`Categorical`, :class:`DatetimeArray`, :class:`FloatingArray`, :class:`IntegerArray`, :class:`PeriodArray`, :class:`TimedeltaArray`, and :class:`PandasArray` are now keyword-only methods (:issue:`37541`)
 
+ToLatex
+^^^^^^^
+
+- Can now substitute simple fstrings in place of callable functions in to_latex
+
+.. ipython:: python
+    data = [[1, 2, 3], [4, 5, 6], [7, 8, 9.001]]
```
Comment (dsaxton) [inline doc/source/whatsnew/v1.2.0.rst:556]: Missing space here may be causing the CI failure. Also I think this should probably be in the I/O section.

Comment (wlgranados) [reply doc/source/whatsnew/v1.2.0.rst]: Got it thanks!

File: pandas/core/generic.py [hunk 2]
```diff
@@ -3034,7 +3034,7 @@ def to_latex(
             Write row names (index).
         na_rep : str, default 'NaN'
             Missing data representation.
-        formatters : list of functions or dict of {{str: function}}, optional
+        formatters : list of functions/str or dict of {{str: function}}, optional
```
Comment (jreback) [inline pandas/core/generic.py:3037]: can you add a versionchanged 1.2 tag here

File: pandas/core/generic.py [hunk 3]
```diff
@@ -3136,6 +3136,22 @@ def to_latex(
         if multirow is None:
             multirow = config.get_option("display.latex.multirow")
 
+        if is_list_like(formatters) and not isinstance(formatters, dict):
```
Comment (jreback) [inline pandas/core/generic.py:3139]: can you add an Example that uses this (e.g. similar to your test is fine)

File: pandas/core/generic.py [hunk 4]
```diff
@@ -3136,6 +3136,22 @@ def to_latex(
...
+            formatter_elems_type = all(
+                isinstance(elem, str) or callable(elem) for elem in formatters
```
Comment (jreback) [inline pandas/core/generic.py:3141]: can you move *all* of this logic inside the DataFrameFormatter itself

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
cc @ivanovmg if you'd have a look

File: pandas/core/generic.py [hunk 5]
```diff
@@ -3136,6 +3136,22 @@ def to_latex(
...
+            )
+            if formatter_elems_type:
+                formatters = [
+                    (lambda style: lambda x: "{0:{1}}".format(x, style))(style)
```
Comment (ivanovmg) [inline pandas/core/generic.py:3145]: I would suggest not to use two folded lambda functions. This is not easily readable. Can it be replaced with an f-string?

**Review by ivanovmg — CHANGES_REQUESTED** [review]
I agree with Jeff on moving the formatters logic into DataFrameFormatter.
There is a method ``_initialize_formatters``, which can be updated.
However, it will alter formatters not only for latex output, but for others as well (to_string, to_html, to_csv).
Is that ok?

**Comment by jreback** [comment]
> I agree with Jeff on moving the formatters logic into DataFrameFormatter.
> 
> There is a method ``_initialize_formatters``, which can be updated.
> 
> However, it will alter formatters not only for latex output, but for others as well (to_string, to_html, to_csv).
> 
> Is that ok?

i think this is ok but we should have some additional tests in place

File: pandas/core/generic.py (hunk 5 above)
Comment (wlgranados) [reply pandas/core/generic.py:3145]: Initially they were simpler lambda functions to take advantage of the existing callable implementation, but there was a python memory specific error where it created the first anonymous function and replicated it multiple times. 

I believe this is using fstrings? At least for the substitution.  But yeah I agree it's a little hard to read.

File: pandas/core/generic.py (hunk 3 above)
Comment (wlgranados) [reply pandas/core/generic.py]: I was having an issue providing an example with lambda functions in doc, so I included  a simpler example.

**Comment by wlgranados** [comment]
> > I agree with Jeff on moving the formatters logic into DataFrameFormatter.
> > There is a method `_initialize_formatters`, which can be updated.
> > However, it will alter formatters not only for latex output, but for others as well (to_string, to_html, to_csv).
> > Is that ok?
> 
> i think this is ok but we should have some additional tests in place

Which tests did you have in mind? I think the FormattersType change looks self contained with the DataFrameFormatter class

File: pandas/io/formats/format.py [hunk 6]
```diff
@@ -576,6 +580,24 @@ def _initialize_sparsify(self, sparsify: Optional[bool]) -> bool:
     def _initialize_formatters(
         self, formatters: Optional[FormattersType]
     ) -> FormattersType:
+        if is_list_like(formatters) and not isinstance(formatters, dict):
+            formatter_elems_type = all(
+                isinstance(elem, str) or callable(elem) for elem in formatters
```
[the remark this answers is not in the document]
Comment (ivanovmg) [reply pandas/io/formats/format.py:585]: Probably should look like this:
```
if formatters is None:
    return {}
elif len(self.frame.columns) == len(formatters) or isinstance(formatters, dict):
    return formatters
elif is_list_like(formatters) and not isinstance(formatters, dict):
    ... and so on
```

Also raising ValueError should be in agreement (probably need to update the error message on what formatter elements should look like).

**Review by ivanovmg — CHANGES_REQUESTED** [review]
See comments

File: pandas/io/formats/format.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wlgranados) [reply pandas/io/formats/format.py:585]: With the ordering you proposed that second if statement would always be evaluated true and it wouldn't convert the fstrings to corresponding functions.  

I'm not too sure how to fix the error, been looking over it slowly past few days. I think mypy is treating the two-folded lambda function as a different type (Like ``List[Callable[... some params]``) than the intended ``List[Callable]``. Kinda odd that it didn't have this issue before though 🤔

**Review by arw2019 — COMMENTED** [review]
@wlgranados how's this going? If you're interested in continuing please merge master to see if you can get to green and ping when ready for another round of reviews

**Comment by jreback** [comment]
closing as stale. if you want to continue, pls ping and can re-open.

**Pull request closed without merging**